### PR TITLE
Swap build side for outer joins when natural build side is explosive [databricks]

### LINKED ISSUE #11234: Explore swapping build table for left outer joins
**Is your feature request related to a problem? Please describe.**
When performing a left outer join on a small left input and a large right input, we're going to use the large input to build the hash table and the small input to probe it.  This could lead to poor performance on the GPU because there's potentially many collisions during the build and potentially lower parallelism on the probe side after the build.

**Describe the solution you'd like**
In this scenario, we could use the much smaller left input as the hash table, but then we need to solve the problem of identifying rows in the left input that were not "hit' during the probe from the right input.  In this case we could take a similar approach that we do for full outer joins, which is use the left input gather map as a scatter map for false to build a filter mask identifying which rows were hit during the probe, and the remaining rows after the filter are the ones not matched to any row in the right input.

### PR DESCRIPTION
**Pull request by jlowe**
Fixes #11234.  Adds a new asymmetric hash join exec that is similar to the existing symmetric hash join handling of full outer joins except it has a slightly different algorithm for determining the build side.  It tries the natural build side first.  If that does not fit in a single batch, it will switch to the stream side if that does.  If the build side fits but appears to be explosive due to highly replicated keys, it will pivot to the stream side if that also fits in a single batch and is less explosive.

### CODE REVIEW

**Comment by jlowe** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala [hunk 1]
```diff
@@ -70,8 +70,24 @@ class GpuShuffledHashJoinMeta(
...
+          joinCondition,
+          left,
+          right,
+          conf.isGPUShuffle,
+          conf.gpuTargetBatchSizeBytes,
+          isSkewJoin = false)(
+          join.leftKeys,
+          join.rightKeys,
+          conf.joinOuterMagnificationThreshold)
+      case Inner | FullOuter if useSizedJoin =>
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:90]: The config changed here. Is that intentional? (I assume that it is)

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala [hunk 2]
```diff
@@ -665,11 +665,29 @@ val GPU_COREDUMP_PIPE_PATTERN = conf("spark.rapids.gpu.coreDump.pipePattern")
...
+      .booleanConf
+      .createWithDefault(true)
+
+  val JOIN_OUTER_MAGNIFICATION_THRESHOLD =
+    conf("spark.rapids.sql.join.outer.magnificationFactorThreshold")
+      .doc("The magnification factor threshold at which outer joins will consider using the " +
+        "unnatural side of the join to build the hash table")
+      .internal()
+      .integerConf
+      .createWithDefault(10000)
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:689]: How did we come up with 10,000 for this?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala [hunk 3]
```diff
@@ -256,9 +256,35 @@ object GpuHashJoin {
...
 
+/**
+ * Class to hold statistics on the build-side batch of a hash join.
+ * @param streamMagnificationFactor estimated magnification of a stream batch during join
+ * @param isDistinct true if all build-side join keys are distinct
+ */
+case class JoinBuildSideStats(streamMagnificationFactor: Double, isDistinct: Boolean)
+
+object JoinBuildSideStats {
+  // TODO: Make boundBuildKeys Seq[GpuExpression]
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:267]: Is this TODO needed?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala [hunk 4]
```diff
@@ -611,48 +627,65 @@ class HashFullJoinStreamSideIterator(
...
       streamAttributes,
       targetSize,
-      FullOuter,
+      joinType,
       buildSide,
       opTime = opTime,
       joinTime = joinTime) {
-  // Full Join is implemented via LeftOuter or RightOuter join, depending on the build side.
-  private val useLeftOuterJoin = (buildSide == GpuBuildRight)
+  private val subJoinType = joinType match {
```
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:639]: nit: Can we have a comment here explaining a bit how this is picked?

**Review by revans2 — COMMENTED** [review]
In general it looks good, but I am a little concerned about some of the tests and databricks. It would be nice to make sure that the tests are run on databrciks too.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala (hunk 1 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffledHashJoinExec.scala:90]: Not sure what you mean by "config changed here."  The old code leveraged conf.useShuffledSymmetricHashJoin for inner and fullouter joins, and the new code is doing the same in  GpuShuffledSizedHashJoinExec.useSizedJoin.  Or did I miss something?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala (hunk 2 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RapidsConf.scala:689]: I did this semi-empirically by creating a build-side table with the same hash key and seeing when it became noticeably slow to use it as the build-side table.  It's not super-scientific, open to hear alternate approaches to determine excessive "explosiveness."

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala (hunk 3 above)
Comment (jlowe) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:267]: Ah, I meant to fix this up and forgot.  It makes the API more type-safe, so I'd like to clean that up.

**Comment by jlowe** [comment]
build

**Pull request merged**